e2e/test/helpers/TestModule.cs

```diff
@@ -3,61 +3,54 @@
-        {
-            _module = module;
-        }
+        /// <summary>
+        /// Used in conjunction with ModuleClient.CreateFromConnectionString()
+        /// </summary>
+        public string ConnectionString =>
+            $"HostName={TestConfiguration.IotHub.GetIotHubHostName()};DeviceId={_module.DeviceId};ModuleId={_module.Id};SharedAccessKey={_module.Authentication.SymmetricKey.PrimaryKey}";
+
+        /// <summary>
+        /// Module Id
+        /// </summary>
+        public string Id => _module.Id;
+
+        /// <summary>
+        /// Device Id
+        /// </summary>
+        public string DeviceId => _testDevice.Id;
 
         /// <summary>
         /// Factory method.
         /// </summary>
         public static async Task<TestModule> GetTestModuleAsync(string deviceNamePrefix, string moduleNamePrefix)
         {
-            using TestDevice testDevice = await TestDevice.GetTestDeviceAsync(deviceNamePrefix).ConfigureAwait(false);
-
-            string deviceName = testDevice.Id;
-            string moduleName = "E2E_" + moduleNamePrefix + Guid.NewGuid();
-
-            using var sc = new IotHubServiceClient(TestConfiguration.IotHub.ConnectionString);
-            VerboseTestLogger.WriteLine($"{nameof(GetTestModuleAsync)}: Creating module for device {deviceName}.");
+            var testModule = new TestModule
+            {
+                _testDevice = await TestDevice.GetTestDeviceAsync(deviceNamePrefix).ConfigureAwait(false)
+            };
 
-            var requestModule = new Module(deviceName, moduleName);
-            Module module = await sc.Modules.CreateAsync(requestModule).ConfigureAwait(false);
+            string deviceName = testModule._testDevice.Id;
+            string moduleName = $"E2E_{moduleNamePrefix}_{Guid.NewGuid()}";
 
-            var ret = new TestModule(module);
+            IotHubServiceClient sc = TestDevice.ServiceClient;
+            testModule._module = await sc.Modules.CreateAsync(new Module(deviceName, moduleName)).ConfigureAwait(false);
+            VerboseTestLogger.WriteLine($"{nameof(GetTestModuleAsync)}: Using device {testModule.DeviceId} with module {testModule.Id}.");
 
-            VerboseTestLogger.WriteLine($"{nameof(GetTestModuleAsync)}: Using device {ret.DeviceId} with module {ret.Id}.");
-            return ret;
+            return testModule;
         }
 
-        /// <summary>
-        /// Used in conjunction with ModuleClient.CreateFromConnectionString()
-        /// </summary>
-        public string ConnectionString
+        public async ValueTask DisposeAsync()
         {
-            get
-            {
-                string iotHubHostName = GetHostName(TestConfiguration.IotHub.ConnectionString);
-                return $"HostName={iotHubHostName};DeviceId={_module.DeviceId};ModuleId={_module.Id};SharedAccessKey={_module.Authentication.SymmetricKey.PrimaryKey}";
-            }
+            await _testDevice.DisposeAsync().ConfigureAwait(false);
```

Comment:
small nit: we could add a comment here stating that TestDevice.DisposeAsync() does in fact delete the associated device as well (since the deleteion code is in a different file).